Fixes from @apt1002, plus some other bug fixes

rrthomas opened the pull request.

Commit:
Use `is_aligned()`, not `smite_is_aligned()`, where possible.

Commit:
Use smite_WORD_BIT, not smite_word_bit, where possible.

rrthomas committed:
src/gen-main.in: use isinstance instead of type (thanks, @apt1002)

rrthomas committed:
doc/smite.tex.in: some clarifications from @apt1002

rrthomas committed:
src/instruction_gen.py: represent stack sizes numerically

Add a Size type, which has a ‘count’ component for variadic sizes.

Use this to generate code precisely and clearly, obviating the need to wrap
some code in compiler warnings.

apt1002 committed:
Break a cyclic build dependency (type_sizes, instruction_gen)

This should probably be done differently.

apt1002 committed:
Cherry-pick @rrthomas's class Size, with a few minor fixes and improvements.

apt1002 committed:
Fixup.

apt1002 committed:
Misc other minor fixes and improvements from instrumented_apt.

apt1002 committed:
Fix: "smite --trace" requires an argument.

apt1002 committed:
Use "in instructions" not "in instructions.__members__.items()".

apt1002 committed:
instruction_gen.py: check_underflow(): Add brackets so the cast has the right precedence.

apt1002 committed:
instruction_gen.py: load_item, store_item: generate code with fewer casts.

apt1002 committed:
Add a file to record the compatible version of pForth

This took me a while to work out!

rrthomas committed:
Simplify specification and implementation of instruction fetch

Thanks @apt1002, for noting a nit that led me to reconsider.

Move all the instruction fetch semantics to the main loop in the spec, where
there is now an if that either shifts I right or loads it from PC.

As a result, NEXT becomes NOP, and BRANCH, BRANCHZ and CALL no longer
perform instruction fetch. This does not affect the value put in BAD when an
error is raised.

This also simplifies the assembler and the branch test.

apt1002 committed:
configure.ac: Configure case fall-through warnings (rescue from instrumented_apt)

apt1002 committed:
src/Makefile.am: pass CPPFLAGS to type-sizes, not CFLAGS

apt1002 committed:
smite_core/instruction_gen.py: fix Size.of, add __neg__

apt1002 committed:
smite_core/instruction_gen.py: add StackItem.of; move load and store from StackEffect to StackItem

apt1002 committed:
src/Makefile.am: exit with error if a generated file cannot be made

apt1002 committed:
Reintroduce class StackPicture

apt1002 committed:
instruction_gen.py: clarify check_{under,over}flow

apt1002 committed:
smite_core/instruction_gen.py: tweak documentation

apt1002 committed:
smite.h: allow compiler to decide what to inline

Remove "inline" keywords, and replace with pragmas so that GCC does
not complain about unused functions.

This prevents warnings about the functions not being inlined when
they have many callers, as for example when a specialized
interpreter is compiled.

rrthomas committed:
Use CPPFLAGS, not CFLAGS, to set WORD_BYTES and ENDISM

rrthomas committed:
Fix implementation of features

Needed more testing!

Split init_code into init_code and parse_code.

rrthomas committed:
Revert "smite_core/instruction_gen.py: update S->STACK_DEPTH only after action routine"

This reverts commit 3960fd4a279e2cd58fbf5613abe1e39eb5ac571d.

The change did not work properly (setting STACK_DEPTH did not work, which
showed up in pForth exception handling), conferred minimal performance
benefit, and wasn’t compatible with the specializer.

apt1002 commented:
I'd like to approve everything except the change to instruction fetch, which I think needs a bit more thought.

My concerns about instruction fetch are:
 - The assembler has regressed. If INSTRUCTION_BITS is 6, say, then the assembler cannot squeeze a low-numbered opcode into the last two bits of a word. `effective_pc()` is unimplementable for this instruction size.
 - There is an error in the specification.
 - The trace file no longer records when instruction fetch happens, and can't.

I'd like to spprove the other specification corrections in the same commit.

rrthomas committed:
Rewrite LibSMite accessor functions without StackEffects

They did work with Stackeffects, but only by coincidence, in particular
manipulating STACK_DEPTH. Do the honest thing and handle the stack manually
instead.

Also add __repr__ method to StackItem, reinstate push and pop macros for
arbitrary C types in smite.h, and clarify Library.types.

apt1002 committed:
instruction_gen.py: dispatch(): Use `append()` rather than +=

apt1002 committed:
Merge branch 'size' into merge-apt-fixes

apt1002 committed:
smite_core/ext.py: make Library handle stack manually

It is a bit of a mystery how this got the way it was. But
something to do with fiddling about with how STACK_DEPTH is
updated.

apt1002 committed:
Revert "Simplify specification and implementation of instruction fetch"

This reverts commit b8029afed6b50ccd54ef28285a52ff16fa045f85, except that
I retained some corrections to "doc/smite.pdf".

apt1002 committed:
doc/smite.tex.in: fix and improve specification

1. Fix a bug: shift I right before executing the instruction.

2. Specify the desired value of BAD consistently in the Errors section.

apt1002 committed:
Python: improve auto_NEXT facility of step()

Auto-run NEXT instructions before rather than after other instructions.

This simplifies some of the tests.

apt1002 reviewed (approved):
Reviewed together.
